### PR TITLE
Allow materialized views to by synchronous

### PR DESCRIPTION
**Pull request by psarna**
This pull request introduces a "synchronous mode" for global views. In this mode, all view updates are applied synchronously as if the view was local.

Marking view as a synchronous one can be done using `CREATE MATERIALIZED VIEW` and `ALTER MATERIALIZED VIEW`. E.g.:
```cql
ALTER MATERIALIZED VIEW ks.v WITH synchronous_updates = true;
```

Marking view as a synchronous one was done using tags (originally used by alternator). No big modifications in the view's code were needed.

Fixes: https://github.com/scylladb/scylla/issues/10545

### CODE REVIEW

**Comment by psarna** [comment]
This is a new version of #10777, because its author is on a long leave

File: .gitignore [hunk 1]
```diff
@@ -22,6 +22,7 @@ resources
 .pytest_cache
 /expressions.tokens
 tags
+!db/tags/
```
Comment (avikivity) [inline .gitignore:25]: What's this?

**Comment by avikivity** [comment]
I don't see doc updates.

I haven't looked deeply, but what's the new tags concept? Why do we need a new concept? Should it not be a schema change?

File: test/cql-pytest/test_materialized_view.py [hunk 2]
```diff
@@ -227,3 +226,31 @@ def test_is_not_operator_must_be_null(cql, test_keyspace):
...
+                cql, table, '*', 'v, p', 'v is not null and p is not null'
+            )
+
+        with new_view() as sync_mv, new_view() as async_mv:
+            # Make one view synchronous
+            cql.execute(f"ALTER MATERIALIZED VIEW {sync_mv} WITH synchronous_updates = true")
+
+            # Execute a query and inspect its tracing info
+            res = cql.execute(f"INSERT INTO {table} (p,v) VALUES (123, 'dog')", trace=True)
+            trace = res.get_query_trace()
```
Comment (nyh) [inline test/cql-pytest/test_materialized_view.py:246]: please check if this request doesn't generate a warning. I think you need to use res.one().get_query_trace()?

File: test/cql-pytest/test_materialized_view.py [hunk 3]
```diff
@@ -227,3 +226,31 @@ def test_is_not_operator_must_be_null(cql, test_keyspace):
...
+# the synchronous updates code path is then reached for such view.
+def test_mv_synchronous_updates(cql, test_keyspace):
+    schema = 'p int, v text, primary key (p)'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        def new_view():
+            return new_materialized_view(
+                cql, table, '*', 'v, p', 'v is not null and p is not null'
+            )
+
+        with new_view() as sync_mv, new_view() as async_mv:
```
Comment (nyh) [inline test/cql-pytest/test_materialized_view.py:240]: Please also create a third view which has synchronous view enabled right at CREATE TABLE time, to check that code path as well.

**Review by nyh — APPROVED** [review]
Looks good. I left a couple of minor comments.

File: cql3/statements/cf_prop_defs.cc [hunk 4]
```diff
@@ -336,6 +337,15 @@ void cf_prop_defs::apply_to_builder(schema_builder& builder, schema::extensions_
     }
 
     builder.set_extensions(std::move(schema_extensions));
+
+    if (has_property(KW_SYNCHRONOUS_UPDATES)) {
```
[the remark this answers is not in the document]
Comment (psarna) [reply cql3/statements/cf_prop_defs.cc:341]: Yes, it's common code for CREATE and ALTER

File: test/cql-pytest/test_materialized_view.py (hunk 2 above)
Comment (psarna) [reply test/cql-pytest/test_materialized_view.py:246]: indeed - I'll fix it in v2

**Comment by psarna** [comment]
v3:
 - extended the test cases for passing `synchronous_updates` in CREATE MATERIALIZED VIEW statement

File: test/cql-pytest/test_materialized_view.py [hunk 5]
```diff
@@ -232,25 +232,28 @@ def test_is_not_operator_must_be_null(cql, test_keyspace):
...
+            unwanted_trace1 = f"Forcing {async_mv} view update to be synchronous"
+            unwanted_trace2 = f"Forcing {async_mv_altered} view update to be synchronous"
 
             wanted_trace_was_found = False
             for event in trace.events:
-                assert unwanted_trace not in event.description
-                if wanted_trace in event.description:
+                assert unwanted_trace1 not in event.description
+                assert unwanted_trace2 not in event.description
+                if wanted_trace1 in event.description or wanted_trace2 in event.description:
```
Comment (psarna) [inline test/cql-pytest/test_materialized_view.py:257]: oh wait, I want both here, not just one -.- will fix in v3

**Comment by psarna** [comment]
v4:
- fixed the test condition to check for both tracing logs

**Comment by slivne** [comment]
How is this applied to secondary indexes  -  can I request a secondary
index with synchronous updates ?

**Comment by psarna** [comment]
> How is this applied to secondary indexes  -  can I request a secondary
index with synchronous updates ?

You can request a secondary index to be synchronous by simply tagging its underlying materialized view to be synchronous. There's no direct syntax to mark an index as synchronous, but it will effectively work just the same. We can consider inventing CQL syntactic sugar to "mark an index synchronous" as a follow-up, to make it more user-friendly.

**Comment by nyh** [comment]
> > How is this applied to secondary indexes  -  can I request a secondary
> > index with synchronous updates ?
> 
> You can request a secondary index to be synchronous by simply tagging its underlying materialized view to be synchronous. There's no direct syntax to mark an index as synchronous, but it will effectively work just the same. We can consider inventing CQL syntactic sugar to "mark an index synchronous" as a follow-up, to make it more user-friendly.

This would make a nice addition to the doc. Users may not think this is possible, and may also not be aware of how the name of the underlying materialized view.

**Review by nyh — APPROVED** [review]
Looks good. Please resolve the conflict in the document (and it's also an opportunity to add a paragraph about secondary index there).

**Comment by psarna** [comment]
v5:
 - added a docs entry on marking global indexes as synchronous

**Pull request merged**